python/lsst/daf/butler/registry/interfaces/database.py

```diff
@@ -0,0 +1,927 @@
+                # how many rows we get back.
+                toSelect.add(next(iter(keys.keys())))
+            selectSql = sqlalchemy.sql.select(
+                [table.columns[k].label(k) for k in toSelect]
+            ).select_from(table).where(
+                sqlalchemy.sql.and_(*[table.columns[k] == v for k, v in keys.items()])
+            )
+            fetched = list(self._connection.execute(selectSql).fetchall())
+            if len(fetched) != 1:
+                return len(fetched), None, None
+            existing = fetched[0]
+            if compared is not None:
+                inconsistencies = [k for k, v in compared.items() if existing[k] != v]
+            else:
+                inconsistencies = []
+            if returning is not None:
+                toReturn = [existing[k] for k in returning]
+            else:
+                toReturn = None
+            return 1, inconsistencies, toReturn
+
+        if self.isWriteable():
+            # Database is writeable.  Try an insert first, but allow it to fail
+            # (in only specific ways).
+            row = keys.copy()
+            if compared is not None:
+                row.update(compared)
+            if extra is not None:
+                row.update(extra)
+            insertSql = table.insert().values(row)
+            try:
+                with self.transaction(interrupting=True):
+                    self._connection.execute(insertSql)
+                    # Need to perform check() for this branch inside the
+                    # transaction, so we roll back an insert that didn't do
+                    # what we expected.  That limits the extent to which we
+                    # can reduce duplication between this block and the other
+                    # ones that perform similar logic.
+                    n, bad, result = check()
+                    if n < 1:
+                        raise RuntimeError("Insertion in sync did not seem to affect table.  This is a bug.")
+                    elif n > 2:
+                        raise RuntimeError("Keys passed to sync do not comprise a unique constraint.")
+                    elif bad:
+                        raise RuntimeError("Conflict in sync after successful insert; this should only be "
+                                           "possible if the same table is being updated by a concurrent "
+                                           "process that isn't using sync.")
+                # No exceptions, so it looks like we inserted the requested row
+                # successfully.
+                inserted = True
+            except sqlalchemy.exc.IntegrityError as err:
+                # Most likely cause is that an equivalent row already exists,
+                # but it could also be some other constraint.  Query for the
+                # row we think we matched to resolve that question.
+                n, bad, result = check()
```

Comment:
Do you need `self.transaction()` around this line?

Reply:
Well, this code should execute outside any transaction (code above uses interrupting=True), and if you don't manage transactions yourself then it is probably sqlalchemy auto-commit feature that handles transactions for you (https://docs.sqlalchemy.org/en/13/core/connections.html#understanding-autocommit). I think it's better to manage transactions explicitly, especially if you tracking active transactions in your class.